// ualib/lib/ua_string.js

var fs = require('fs');

function get_xml_value(str, s_tag, e_tag) {
	var p = -1;
	var start = -1;
	var end = -1;
	var val = null;

	p = str.indexOf(s_tag);
	if (p >= 0 ) {
		start = p + s_tag.length;
		end = str.indexOf(e_tag, start);
		val = str.substring(start, end);
		return val;
	}
	else {
		return null;
	}
}

function get_xml_value_obj(str, s_tag, e_tag, point) {
	var p = -1;
	var start = -1;
	var end = -1;
	var val = null;
	var ret = {};

	if (point <= -1) {
		point = 0;
	}

	p = str.indexOf(s_tag, point);
	if (p >= 0 ) {
		start = p + s_tag.length;
		end = str.indexOf(e_tag, start);
		val = str.substring(start, end);

		ret['val'] = val;
		ret['end'] = end + e_tag.length;

		return ret;
	}
	else {
		ret['val'] = null;
		ret['end'] = -1;

		return ret;
	}
}

function base64_encode(file) {
	var bitmap = fs.readFileSync(file);
	return new Buffer(bitmap).toString('base64');
}

function base64_decode(base64str, file) {
	var bitmap = new Buffer(base64str, 'base64');
	fs.writeFileSync(file, bitmap);
}

exports.get_xml_value = get_xml_value;
exports.get_xml_value_obj = get_xml_value_obj;
exports.base64_encode = base64_encode;
exports.base64_decode = base64_decode;

/*
var ret = null;
var data = 	'<div>seo</div>\n' + 
			'<div2>seok</div2>\n' +
			'<div3 div="seok">eun</div3>\n' +
			'<div4 div="seok"></div4>\n';

ret = get_xml_value(data, 'div>', '</div>');
console.log('ret = ' + ret);
ret = get_xml_value(data, 'seok>', '</seok>');
console.log('ret = ' + ret);

ret = get_xml_value_obj(data, 'div>', '</div>', -2);
console.log('ret = ' + JSON.stringify(ret));
console.log('ret.val = ' + ret.val);
ret = get_xml_value_obj(data, 'seok>', '</seok>', -2);
console.log('ret = ' + JSON.stringify(ret));

// convert image to base64 encoded string
var base64str = base64_encode('11.jpg');
console.log(base64str);
fs.writeFileSync('11.jpg.base64', base64str);

// convert base64 string back to image
base64_decode(base64str, 'copy.jpg');
*/
